sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/NaiveSpannerReadFn.java

```diff
@@ -0,0 +1,49 @@
+package org.apache.beam.sdk.io.gcp.spanner;
+
+import com.google.cloud.spanner.ReadOnlyTransaction;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.TimestampBound;
+import com.google.common.annotations.VisibleForTesting;
+
+/** A simplest read function implementation. Parallelism support is coming. */
+@VisibleForTesting
+class NaiveSpannerReadFn extends AbstractSpannerFn<Object, Struct> {
+
+  private final SpannerIO.Read config;
+
+  NaiveSpannerReadFn(SpannerIO.Read config) {
+    this.config = config;
+  }
+
+  SpannerConfig getSpannerConfig() {
+    return config.getSpannerConfig();
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) throws Exception {
+    TimestampBound timestampBound = TimestampBound.strong();
+    if (config.getTransaction() != null) {
+      Transaction transaction = c.sideInput(config.getTransaction());
+      timestampBound = TimestampBound.ofReadTimestamp(transaction.timestamp());
+    }
+    try (ReadOnlyTransaction readOnlyTransaction =
+        databaseClient().readOnlyTransaction(timestampBound)) {
+      ResultSet resultSet = execute(readOnlyTransaction);
+      while (resultSet.next()) {
+        c.output(resultSet.getCurrentRowAsStruct());
```

Comment:
We definitely need something like this.

It is something that we will need to design first, maybe make the api consistent with writes.